ard_controller/ard_controller.ino

```diff
@@ -15,45 +25,274 @@ struct JointAngleStruct {
+          validated_packet_data.left_knee,
+          validated_packet_data.right_hip,
+          validated_packet_data.right_knee,
+          validated_packet_data.left_shoulder,
+          validated_packet_data.left_elbow,
+          validated_packet_data.right_shoulder,
+          validated_packet_data.right_elbow,
+          validated_packet_data.checksum,
+          validated_packet_data.checksum_error,
+          validated_packet_data.packet_available,
+          validated_packet_data.data_request
+          );
+  Serial.println(buffer);
+}
+
+void dump_temporary_packet_data() {
+  sprintf(buffer, "%3u %3u %3u %3u %3u %3u %3u %3u %3u %3u %3u %3u",
+          temporary_packet_data.left_hip,
+          temporary_packet_data.left_knee,
+          temporary_packet_data.right_hip,
+          temporary_packet_data.right_knee,
+          temporary_packet_data.left_shoulder,
+          temporary_packet_data.left_elbow,
+          temporary_packet_data.right_shoulder,
+          temporary_packet_data.right_elbow,
+          temporary_packet_data.checksum,
+          temporary_packet_data.checksum_error,
+          temporary_packet_data.packet_available,
+          temporary_packet_data.data_request
+          );
+  Serial.println(buffer);
+}
+
 void rx_processor() {
   static int preamble_counter = PREAMBLE_LENGTH;   // Number of ints in the packet preamble
+  static int data_byte_counter = DATA_BYTE_LENGTH; // Number of ints per joint angle
   static int calculated_checksum = 0xFF; // We have to keep a running checksum of incoming data to verify validity
   int received_data; // The current int read from the RPi
 
   if (Serial1.available() > 0) {
-    received_data = Serial1.read();
+    if (sr_state != INIT) {
+      received_data = Serial1.read();
+      // Serial.println(received_data);
+    }
+
+    switch (sr_state)
+    {
+      case INIT:
+        calculated_checksum = 0x00;
+        preamble_counter = PREAMBLE_LENGTH;
+
+        temporary_packet_data.left_hip = 0;
+        temporary_packet_data.left_knee = 0;
+        temporary_packet_data.right_hip = 0;
+        temporary_packet_data.right_knee = 0;
+        temporary_packet_data.left_shoulder = 0;
+        temporary_packet_data.left_elbow = 0;
+        temporary_packet_data.right_shoulder = 0;
+        temporary_packet_data.right_elbow = 0;
+        temporary_packet_data.checksum = 0;
+        temporary_packet_data.checksum_error = false;
+        temporary_packet_data.packet_available = false;
+        temporary_packet_data.data_request = false;
+        sr_state = READ_PREAMBLE;
+        break;
+      case READ_PREAMBLE:
+        if (received_data == 0xFF) {
+          preamble_counter--;
+          if (preamble_counter == 0) {
+            preamble_counter = PREAMBLE_LENGTH;
+            sr_state = READ_L_HIP;
+          }
+        }
+        else {
+          preamble_counter = PREAMBLE_LENGTH;
+        }
+        break;
+      case READ_L_HIP:
```

Comment:
How do we differentiate between front and rear joints?
I think we should have a differentiator in the variable name

Reply:
There already is a differentiator. the rear joints are labeled "hip" and "knee" while the front are labeled "shoulder" and "elbow"

Reply:
Awesome!! Did not catch that